mantidproject/mantid
# Add calculation to invert a symmetric tridiagonal matrix and speed up MonteCarloAbsorption

**Description of work.**

The error calculation in HistogramData::interpolateYCSplineInplace requires a matrix to be inverted. This has caused the absorption calculations in MonteCarloAbsorption to slow down in some cases when spline interpolation is used (mainly when a sparse instrument calculation is run) on a workspace with a large number of bins
This change speeds up the absorption calculations by implementing a fast analytical calculation to invert a tridiagonal matrix in the special case where it's symmetric and the diagonal elements (in each of the 3 diagonals) are the same. If the matrix doesn't satisfy these conditions, the inversion function just reverts to the standard Matrix::Invert function.
The analytic calculation is described here:
"Analytical inversion of symmetric tridiagonal matrices", G Y Hu and R F O'Connell, J. Phys. A: Math. Gen. 29 (1996)
[AnalyticalInversionOfSymmetricTridiagonalMatrices.pdf](https://github.com/mantidproject/mantid/files/5163931/AnalyticalInversionOfSymmetricTridiagonalMatrices.pdf)

Having worked through the derivation in the paper I got a couple of different minus signs for the case where -2 < D < 2. So my implementation isn't a 100% match. My implementation does produce correct inverses on simple 2x2 matrices though with D values in this range so happy these extra features in my code are correct


**To test:**

Run ILLDirectGeometryReduction.IN4 system test (on release build) before and after this PR. This test runs the DirectILLSelfShielding algorithm which in turn runs the MonteCarloAbsorption algorithm with a sparse instrument with CSpline interpolation specified. When run locally I see a speed up from about 1m 10 seconds to about 50 seconds

Fixes #29253 

---

#### Reviewer ####

Please comment on the following ([full description](http://developer.mantidproject.org/ReviewingAPullRequest.html)):

##### Code Review #####

- Is the code of an acceptable quality?
- Does the code conform to the [coding standards](http://developer.mantidproject.org/Standards/)?
- Are the unit tests small and test the class in isolation?
- If there is GUI work does it follow the [GUI standards](http://developer.mantidproject.org/Standards/GUIStandards.html)?
- If there are changes in the release notes then do they describe the changes appropriately?

##### Functional Tests #####

- Do changes function as described? Add comments below that describe the tests performed?
- Do the changes handle unexpected situations, e.g. bad input?
- Has the relevant (user and developer) documentation been added/updated?

Does everything look good? Mark the review as **Approve**. A member of `@mantidproject/gatekeepers` will take care of it.
